tesseract-server/src/handlers/logic_layer/aggregate.rs

```diff
@@ -206,3 +332,682 @@ pub fn logic_layer_aggregation(
+    let mut added_drilldowns: Vec<LevelName> = vec![];
+
+    // Populate the vectors above
+    for (_dimension_name, level_cuts_map) in dimension_cuts_map.iter() {
+        let mut inner_cuts: Vec<Cut> = vec![];
+
+        let num_level_cuts = level_cuts_map.len();
+
+        for (level_name, level_cuts) in level_cuts_map.iter() {
+            let cut = Cut {
+                level_name: level_name.clone(),
+                members: level_cuts.clone(),
+                mask: Mask::Include,
+                for_match: false
+            };
+
+            inner_cuts.push(cut.clone());
+
+            if num_level_cuts > 1 {
+                // We're doing multiple cuts on this dimension
+                added_drilldowns.push(cut.level_name.clone());
+            }
+        }
+
+        dimension_cuts.push(inner_cuts);
+    }
+
+    // All the different TsQuery's that need to be performed
+    let mut queries: Vec<TsQuery> = vec![];
+
+    // Get all possible combinations of cuts across dimensions
+    let cut_combinations: Vec<Vec<Cut>> = cartesian_product(dimension_cuts);
+
+    // Create a TsQuery for each cut combination
+    for cut_combination in &cut_combinations {
+        let mut drills = drilldowns.clone();
+        let mut caps = captions.clone();
+
+        for cut in cut_combination.clone() {
+            // Look for drilldowns that might need to be added
+            if added_drilldowns.contains(&cut.level_name) {
+                drills.push(Drilldown(cut.level_name.clone()));
+
+                let level = match cube.get_level(&cut.level_name) {
+                    Some(level) => level,
+                    None => break
+                };
+
+                // Add captions for this level
+                let new_captions = level.get_captions(&cut.level_name, &locales);
+                caps.extend_from_slice(&new_captions);
+            }
+        }
+
+        // Populate queries vector
+        queries.push(TsQuery {
+            drilldowns: drills,
+            cuts: cut_combination.clone(),
+            measures: measures.clone(),
+            parents: parents.clone(),
+            properties: properties.clone(),
+            captions: caps,
+            top: top.clone(),
+            top_where: top_where.clone(),
+            sort: sort.clone(),
+            limit: limit.clone(),
+            rca: rca.clone(),
+            growth: growth.clone(),
+            debug: debug.clone(),
+            exclude_default_members: exclude_default_members.clone(),
+            filters: filters.clone(),
+            rate: rate.clone(),
+            sparse: sparse.clone(),
+        });
+    }
+
+    Ok((queries, header_map))
+
+}
+
+
+/// Given a vector containing a partial Cartesian product, and a list of items,
+/// return a vector adding the list of items to the partial Cartesian product.
+/// From: https://gist.github.com/kylewlacy/115965b40e02a3325558
+pub fn partial_cartesian<T: Clone>(a: Vec<Vec<T>>, b: Vec<T>) -> Vec<Vec<T>> {
```

Comment:
can the cartesian fns get put into the util module? Also, just wanted to double check that this functionality can't be provided by itertools: https://docs.rs/itertools/0.8.0/itertools/trait.Itertools.html#method.cartesian_product

Reply:
I don't think these are quite equivalent. The `itertools` method generates pairs between two iterators. We're doing something a little different here.